dref/serializers.py

```diff
@@ -0,0 +1,220 @@
+    created_by_details = UserNameSerializer(source='created_by', read_only=True)
+    users_details = UserNameSerializer(source='users', many=True, read_only=True)
+    budget_file_details = DrefFileSerializer(source='budget_file', read_only=True)
+    cover_image_details = DrefFileSerializer(source='cover_image', read_only=True)
+    disaster_type_details = DisasterTypeSerializer(source='disaster_type', read_only=True)
+
+    class Meta:
+        model = Dref
+        fields = '__all__'
+        read_only_fields = ('modified_by', 'created_by', 'budget_file_preview')
+
+    def remove_digits_after_decimal(self, field):
+        # NOTE: We are doing this to remove decimal after 3 digits whole numbers
+        # eg: 100.00% be replaced with 100%
+        if field and len(field.split('.')[0]) == 3:
+            return field.split('.')[0]
+        return None
+
+    def to_representation(self, instance):
+        data = super().to_representation(instance)
+        disability_people_per = data.get('disability_people_per', '')
+        people_per_urban = data.get('people_per_urban', '')
+        people_per_local = data.get('people_per_local', '')
+        data['disability_people_per'] = self.remove_digits_after_decimal(disability_people_per)
+        data['people_per_urban'] = self.remove_digits_after_decimal(people_per_urban)
+        data['people_per_local'] = self.remove_digits_after_decimal(people_per_local)
+        return data
+
+    def validate(self, data):
+        event_date = data.get('event_date')
+        if event_date and data['type_of_onset'] not in [Dref.OnsetType.SLOW, Dref.OnsetType.SUDDEN]:
+            raise serializers.ValidationError({
+                'event_date': ugettext('Cannot add event_date if onset type not in %s or %s' % (Dref.OnsetType.SLOW.label, Dref.OnsetType.SUDDEN.label))
+            })
+        return data
+
+    def validate_images(self, images):
+        # Validate if images are provided
+        images_id = [image.id for image in images]
+        instance_images = self.instance and self.instance.images.all() or []
+        not_created_by_user_images = DrefFile.objects.filter(
+            # Not created by current user
+            ~models.Q(created_by=self.context['request'].user),
+            # Look into provided images
+            id__in=images_id
+        ).exclude(
+            # Exclude already attached images
+            id__in=instance_images
+        ).values_list('id', flat=True)
+        for image in not_created_by_user_images:
+            raise serializers.ValidationError(
+                ugettext('Only image owner can attach image(id: %s)' % image)
+            )
+        # Don't allow images more than MAX_NUMBER_OF_IMAGES
+        if len(images) > self.MAX_NUMBER_OF_IMAGES:
+            raise serializers.ValidationError(
+                ugettext('Can add utmost %s images' % self.MAX_NUMBER_OF_IMAGES)
+            )
+        return images
```

Comment:
Continuation of https://github.com/IFRCGo/go-api/pull/1141#discussion_r798229277

```suggestion
        # Don't allow images more than MAX_NUMBER_OF_IMAGES
        if len(images) > self.MAX_NUMBER_OF_IMAGES:
            raise serializers.ValidationError(
                ugettext('Can add utmost %s images' % self.MAX_NUMBER_OF_IMAGES)
            )
        images_id = [image.id for image in images]
        images_without_access_qs = DrefFile.objects.filter(
            # Not created by current user
            ~models.Q(created_by=self.context['request'].user),
            # Look into provided images
            id__in=images_id,
        )
        # Exclude already attached images if exists
        if self.instance:
            images_without_access_qs = images_without_access_qs.exclude(
                id__in=self.instance.images.all()
            )
        images_id_without_access = images_without_access_qs.values_list('id', flat=True)
        if images_id_without_access:
            raise serializers.ValidationError(
                ugettext(
                    'Only image owner can attach image. Not allowed image ids: %s' % ','.join(map(str, images_id_without_access))
                )
            )
        return images
```